README
Graph::Timeline - Manipulate and display timelines

This is release V1.4 of Graph::Timeline. The class 
is used to show timelines.

Look in the examples directory for a program showing how to use the
module and some sample data. The first type of chart is the traditional
sort you see in history books showing the lives of famous people against
the timeline of world events.

The second type lists events down the page, starting with the eariest.
This will form a sort of diagonal. Used less for historical data and more
for charting processes.

You can also generate an imagemap for this image

INSTALLATION

 % perl Makefile.PL
 % make
 % make test
 % sudo make install

TEST FAILURES

The tests are there to make sure that I have broken nothing when I
fiddle with the code and will teach you very little about how to use
the code. 

TO DO

* Look at getting 1971/05-1971/09 for sometime between May 1971 and September
  1971 accepted for 'start' and 'end' dates for intervals and points.
* Documentation

BUGS

None so far

REQUIREMENTS

* Developed under 5.6.1 but should work on anything 5+.
* Written completely in Perl. XS is not required.
* Requires Date::Calc, GD, GD::Text

AUTHOR

Peter Hickman <peterhi@ntlworld.com>

Copyright (c) 2003, Peter Hickman. All rights reserved.

This module is free software. It may be used, redistributed and/or 
modified under the same terms as Perl itself.

CHANGES IN THIS RELEASE

V1.0:
      - First release to CPAN

V1.1:
      - Handles the years in the window correctly
	  - Draws points

V1.2:
      - Minor fix to the POD

V1.3:
	  - New DiagonalGD chart type


Changes
Revision history for Perl extension Graph::Timeline

1.4  Wed 20th June 2007
    - New map method for DiagonalGD allow the user to create an
      imagemap.

1.3  Sun 17th June 2007
    - Start and end can now be input down to the second level with
      the yyyy-mm-ddThh:mm:ss format. However GraphGD only operates
      at the yyyy-mm-dd level. The new format was introduced for the
      DiagonalGD chart.

1.2  Wed 14th September 2005
	- Realised I needed a change log file


lib/Graph/Timeline/GD.pm
package Graph::Timeline::GD;

use strict;
use warnings;

use GD;
use GD::Text::Wrap;

use base 'Graph::Timeline';

our $VERSION = '1.4';

sub render {
    die "Timeline::GD->render() expected HASH as parameter" unless scalar(@_) % 2 == 1;

    my ( $self, %data ) = @_;

    %data = $self->_lowercase_keys(%data);
    $self->_valid_keys( 'render', \%data, (qw/border pixelsperday pixelspermonth pixelsperyear/) );
    $data{border} = 0 unless $data{border};

    # Validate the parameters

    my $counter = 0;
    $counter++ if $data{pixelsperday};
    $counter++ if $data{pixelspermonth};
    $counter++ if $data{pixelsperyear};

    if ( $counter == 0 ) {
        die "Timeline::GD->render() one of 'pixelsperday', 'pixelspermonth' or 'pixelsperyear' must be defined";
    }
    elsif ( $counter > 1 ) {
        die "Timeline::GD->render() only one of 'pixelsperday', 'pixelspermonth' or 'pixelsperyear' can be defined";
    }

    # Get the data to render

    my @pool = $self->data();

    die "Timeline::GD->render() there is no data to render" if scalar(@pool) == 2;

    my ( $start, $end ) = $self->_get_start_and_end(@pool);

    # Work out the width of a year in pixels

    my %years;
    my $image_width = 0;

    foreach my $year ( $start .. $end ) {
        $years{$year}->{year} = $year;
        $years{$year}->{days_in_year} = Date::Calc::Days_in_Year( $year, 12 );
        if ( $data{pixelsperday} ) {
            $years{$year}->{pixels_in_year} = $years{$year}->{days_in_year} * $data{pixelsperday};
        }
        elsif ( $data{pixelspermonth} ) {
            $years{$year}->{pixels_in_year} = 12 * $data{pixelspermonth};
        }
        else {
            $years{$year}->{pixels_in_year} = $data{pixelsperyear};
        }

        $image_width += $years{$year}->{pixels_in_year};

        $self->render_year( $years{$year} );
    }

    # Now we should build up the streams for the points and intervals

    my %intervals;
    my $sequence = 1;

    foreach my $record (@pool) {
        if ( $record->{type} eq 'interval' ) {
            my $done = 0;

            my $group = $record->{group};

            $record->{width}      = $self->_calculate_width( $record, 'start_start', 'end_end',   %years );
            $record->{width_pre}  = $self->_calculate_width( $record, 'start_start', 'start_end', %years );
            $record->{width_post} = $self->_calculate_width( $record, 'end_start',   'end_end',   %years );
            $self->render_interval($record);

            foreach my $stream ( @{ $intervals{$group} } ) {
                if ( $stream->[-1]->{end_end} lt $record->{start_start} ) {
                    push( @{$stream}, $record );
                    $done = 1;
                    last;
                }
            }

            push( @{ $intervals{$group} }, [$record] ) unless $done;
        }
        else {
            $record->{sequence} = $sequence++;
            $self->render_point($record);
            push( @{ $intervals{'--points--'}[0] }, $record );
        }
    }

    # Work out the full height of the image

    my $image_height = 0;

    # First the years

    my $max = 0;
    foreach my $year ( $start .. $end ) {
        die "Timeline::GD->render() key 'height' is not defined from render_year()" unless $years{$year}->{height};
        $max = $years{$year}->{height} if $years{$year}->{height} > $max;
    }
    $image_height += $max;

    # Then the intervals

    foreach my $group ( keys %intervals ) {
        foreach my $stream ( @{ $intervals{$group} } ) {
            $max = 0;
            foreach my $entry ( @{$stream} ) {
                die "Timeline::GD->render() key 'height' is not defined from render_" . $entry->{type} . "()" unless $entry->{height};
                $max = $entry->{height} if $entry->{height} > $max;
            }
            $image_height += $max;
        }
    }

    my $im = GD::Image->new( $image_width + ( 2 * $data{border} ), $image_height + ( 2 * $data{border} ) );
    my $white = $im->colorAllocate( 255, 255, 255 );

    my $ypointer = $data{border};

    # Render the big image, points first

    my $xpointer = $data{border};

    $max = 0;

    foreach my $entry ( @{ $intervals{'--points--'}[0] } ) {
        $max = $entry->{height};
        $xpointer = $data{border} + $self->_calc_start_x( $start, $entry->{start}, %years );
        $im->copy( $entry->{data}, $xpointer, $ypointer, 0, 0, $entry->{width}, $entry->{height} );
    }

    $ypointer += $max;

    # Render the big image, years next

    $xpointer = $data{border};

    foreach my $year ( $start .. $end ) {
        $im->copy( $years{$year}->{data}, $xpointer, $ypointer, 0, 0, $years{$year}->{pixels_in_year}, $years{$year}->{height} );
        $xpointer += $years{$year}->{pixels_in_year};
    }

    $ypointer += $years{$start}->{height};

    # Render the big image, intervals last

    foreach my $group ( sort keys %intervals ) {
        if ( $group ne '--points--' ) {
            foreach my $stream ( @{ $intervals{$group} } ) {
                $max = 0;
                foreach my $entry ( @{$stream} ) {
                    $max = $entry->{height} if $entry->{height} > $max;
                    $xpointer = $data{border} + $self->_calc_start_x( $start, $entry->{start_start}, %years );
                    $im->copy( $entry->{data}, $xpointer, $ypointer, 0, 0, $entry->{width}, $entry->{height} );
                }
                $ypointer += $max;
            }
        }
    }

    # Return the data

    return $im->png;
}

sub render_year {
    my ( $self, $year ) = @_;

    # height of a year

    $year->{height} = 15;

    # Create a year line

    my $im = GD::Image->new( $year->{pixels_in_year}, $year->{height} );
    my $base;

    if ( $year->{year} % 2 == 0 ) {
        $base = $im->colorAllocate( 255, 0, 0 );
    }
    else {
        $base = $im->colorAllocate( 0, 255, 0 );
    }

    my $ink = $im->colorAllocate( 255, 255, 255 );

    my $wrapbox = GD::Text::Wrap->new(
        $im,
        line_space => 4,
        color      => $ink,
        text       => $year->{year},
        align      => 'center',
    );

    $wrapbox->set_font(gdSmallFont);

    $wrapbox->draw( 0, 0 );

    $year->{data} = $im;
}

sub render_interval {
    my ( $self, $record ) = @_;

    # height of a year

    my $height = 30;

    # Create a year line

    my $im = GD::Image->new( $record->{width}, $height );
    my $base = $im->colorAllocate( 127, 127, 127 );
    my $ink  = $im->colorAllocate( 255, 255, 255 );
    my $edge = $im->colorAllocate( 180, 180, 180 );

    if ( $record->{width_pre} ) {
        $im->filledRectangle( 0, 0, $record->{width_pre} - 1, $height, $edge );
    }

    if ( $record->{width_post} ) {
        $im->filledRectangle( $record->{width} - $record->{width_post}, 0, $record->{width}, $height, $edge );
    }

    my $wrapbox = GD::Text::Wrap->new(
        $im,
        line_space => 4,
        color      => $ink,
        text       => $record->{label},
        align      => 'center',
    );

    $wrapbox->set_font(gdSmallFont);

    $wrapbox->draw( 0, 0 );

    $record->{data}   = $im;
    $record->{height} = $height;
}

sub render_point {
    my ( $self, $record ) = @_;

    # height and width of a point

    my $height = 30;
    my $width  = 100;

    my $im = GD::Image->new( $width, $height );
    my $base = $im->colorAllocate( 255, 255, 255 );
    my $ink  = $im->colorAllocate( 0,   0,   0 );

    $im->transparent($base);

    my $wrapbox = GD::Text::Wrap->new(
        $im,
        width      => ( $width - 2 ),
        height     => ( $height / 2 ),
        line_space => 4,
        color      => $ink,
        text       => $record->{label},
        align      => 'left',
    );

    $wrapbox->set_font(gdSmallFont);

    if ( $record->{sequence} % 2 == 1 ) {
        $wrapbox->draw( 2, 0 );
        $im->line( 0, 0, 0, $height, $ink );
    }
    else {
        $wrapbox->draw( 2, ( $height / 2 ) );
        $im->line( 0, ( $height / 2 ), 0, $height, $ink );
    }

    $record->{data}   = $im;
    $record->{height} = $height;
    $record->{width}  = $width;
}

sub _calculate_width {
    my ( $self, $record, $start, $end, %years ) = @_;

    return 0 if $record->{$start} eq $record->{$end};

    my ( $first_year, $first_month, $first_day ) = split( '\/', ( split( 'T', $record->{$start} ) )[0] );
    my ( $last_year,  $last_month,  $last_day )  = split( '\/', ( split( 'T', $record->{$end} ) )[0] );

    # Calculate pixel width

    my $width = 0;

    if ( $first_year eq $last_year ) {
        $width += ( $years{$first_year}->{pixels_in_year} / $years{$first_year}->{days_in_year} ) * ( Date::Calc::Delta_Days( $first_year, $first_month, $first_day, $last_year, $last_month, $last_day ) + 1 );
    }
    else {
        foreach my $year ( $first_year .. $last_year ) {
            if ( $year == $first_year ) {
                $width += ( $years{$year}->{pixels_in_year} / $years{$year}->{days_in_year} ) * ( Date::Calc::Delta_Days( $first_year, $first_month, $first_day, $first_year, '12', '31' ) + 1 );
            }
            elsif ( $year == $last_year ) {
                $width += ( $years{$year}->{pixels_in_year} / $years{$year}->{days_in_year} ) * ( Date::Calc::Delta_Days( $last_year, 1, 1, $last_year, $last_month, $last_day ) + 1 );
            }
            else {
                $width += $years{$year}->{pixels_in_year};
            }
        }
    }

    return int($width);
}

sub _calc_start_x {
    my ( $self, $start_graph, $start_interval, %years ) = @_;

    my ( $first_year, $first_month, $first_day ) = split( '\/', ( split( 'T', $start_graph ) )[0] );
    my ( $last_year,  $last_month,  $last_day )  = split( '\/', ( split( 'T', $start_interval ) )[0] );

    my $x = 0;

    foreach my $year ( $first_year .. $last_year ) {
        if ( $year != $last_year ) {
            $x += $years{$year}->{pixels_in_year};
        }
        else {
            $x += ( $years{$year}->{pixels_in_year} / $years{$year}->{days_in_year} ) * ( Date::Calc::Delta_Days( $last_year, 1, 1, $last_year, $last_month, $last_day ) + 1 );
        }
    }

    return $x;
}

sub _get_start_and_end {
    my ( $self, @pool ) = @_;

    my $start = $pool[0]->{start};
    my $end   = $pool[0]->{end};

    foreach my $record (@pool) {
        $end = $record->{end} if $record->{end} gt $end;
    }

    $start = ( split( '\/', $start ) )[0];
    $end   = ( split( '\/', $end ) )[0];

    return $start, $end;
}

1;

=head1 NAME

Graph::Timeline::GD - Render timeline data with GD

=head1 VERSION

This document refers to verion 1.4 of Graph::Timeline::GD, released June 20, 2007

=head1 SYNOPSIS

This subclass produces the GD object of the timeline. The user has to subclass from this class if they want
a GD rendering of the timeline data. By overriding the render_year( ), render_point( ) and render_interval( ) 
methods the user can supply a less garish and more pleasing display.

  use Graph::Timeline::GD;

  my $x = Graph::Timeline::GD->new();

  while ( my $line = <> ) {
    chomp($line);

    my ( $label, $start, $end, $group ) = split ( ',', $line );
    if($end) {
      $x->add_interval( label => $label, start => $start, end => $end, group => $group );
    }
    else {
      $x->add_point( label => $label, start => $start, group => $group );
    }
  }

  $x->window(start=>'1900/01/01', end=>'1999/12/31');

  open(FILE, '>test.png');
  binmode(FILE);
  print FILE $x->render( border => 2, pixelsperyear => 35 );
  close(FILE);

All the user needs to do is create a package that subclasses Graph::Timeline::GD

  package MyTimeLine;

  use base Graph::Timeline::GD;

  sub render_year { ... }

  sub render_interval { ... }

  sub render point { ... }

  1;

The default methods in Graph::Timeline::GD will show you how to write your own methods and the timeline 
script in the examples directory will show you how read in data, set up the timeline and draw various 
graphs with it.

=head1 DESCRIPTION

=head2 Overview

Only three methods need to be overridden to create your own GD image of the data. 

=over 4

=item render_year( YEAR )

The years that form the axis of the graph are rendered by render_year( ). A scalar pointing to the data 
for the year to be rendered is passed to the method. All you have to do is create an image of the correct 
size and decorate it.

=item render_interval( RECORD )

To render an interval this method takes the record of the interval. RECORD is a pointer to a hash that 
contains the all the data you should require, the important ones are:

=over 4

=item width

This the width of the required image.

=item label

The label that came from the data.

=item group

The group that the interval belongs to.

=back

Additionally the following are also defined but you may have no need for them

=over 4

=item end, end_start, end_end, width_post

End is the end date as defined in the data, end_start and end_end define a subinterval that the end of the data occured in.
For example if the end date is 1980/12/15 then end, end_start and end_end will be the same and width_post will be 0. However
should the end date be an interval like 1980/12 (something during December 1980) then end_start will be 1980/12/01 and
end_end will be 1980/12/31. Width_post will contain the number of pixels that represent the width of the subinterval.

=item start, start_start, start_end, width_pre

The same subinterval messing about for the start date as for the end date (defined above).

=back

=item render_point( RECORD )

Just the same as render_interval but with the addition of the sequence data, as points are rendered they
are numbered sequentualy from 1.

=back 

=head2 Constructors and initialisation

=over 4

=item new( )

Inherited from Graph::Timeline

=back

=head2 Public methods

=over 4

=item render( HASH )

The method called to return the rendered image. This takes a hash of configuration options however only
one of the pixelsper* keys can be supplied (being as they are mutually exclusive) and border is optional.

=over 4

=item border

The number of pixels to use as a border around the graph.

=item pixelsperyear

The number of pixels the year will be rendered in

=item pixelspermonth

The number of pixels to render a month in, the number of pixels a year will be this value times twelve

=item pixelsperday

The number of pixels to render a day in, the number of pixels in a year will be calculated from this

=back

=item render_year( SCALAR )

Override this method to render a year.

=item render_interval( SCALAR )

Override this method to render an interval.

=item render_point( SCALAR )

Override this method to render a point.

=back

=head2 Private methods

=over 4

=item _calculate_width

A method to calculate the width in pixels of an interval

=item _calc_start_x

A method to calculate at what offset a year, interval or point should be placed in the final image

=item _get_start_and_end

A method to find the first and last date

=back

=head1 ENVIRONMENT

None

=head1 DIAGNOSTICS

=over 4

=item Timeline->new() takes no arguments

When the constructor is initialised it requires no arguments. This message is given if 
some arguments were supplied.

=item Timeline::GD->render() expected HASH as parameter

Render expects a hash and did not get one

=item Timeline::GD->render() one of 'pixelsperday', 'pixelspermonth' or 'pixelsperyear' must be defined

One of the required parameters needs to be defined

=item Timeline::GD->render() only one of 'pixelsperday', 'pixelspermonth' or 'pixelsperyear' can be defined

Only on parameter can be defined

=item Timeline::GD->render() key 'height' is not defined from render_year()

The method that renders the year has not set the height key, this is required

=item Timeline::GD->render() key 'height' is not defined from render_interval()

The method that renders an interval has not set the height key, this is required

=item Timeline::GD->render() there is no data to render

None of the input data got passed through the call to window()

=back

=head1 BUGS

None

=head1 FILES

See the timeline script in the examples directory

=head1 SEE ALSO

Graph::Timeline - The core timeline class

=head1 AUTHORS

Peter Hickman (peterhi@ntlworld.com)

=head1 COPYRIGHT

Copyright (c) 2003, Peter Hickman. All rights reserved.

This module is free software. It may be used, redistributed and/or 
modified under the same terms as Perl itself.


lib/Graph/Timeline.pm
package Graph::Timeline;

use strict;
use warnings;

use Date::Calc;

our $VERSION = '1.4';

sub new {
    my ($class) = @_;

    die "Timeline->new() takes no arguments" if scalar(@_) != 1;

    my $self = {};

    $self->{_pool} = ();

    return bless $self, $class;
}

sub add_interval {
    die "Timeline->add_interval() expected HASH as parameter" unless scalar(@_) % 2 == 1;

    my ( $self, %data ) = @_;

    %data = $self->_lowercase_keys(%data);
    $self->_required_keys( 'add_interval', \%data, (qw/start end label/) );
    $self->_valid_keys( 'add_interval', \%data, (qw/start end label group id url/) );

    $data{type} = 'interval';

    $self->_add_to_pool(%data);
}

sub add_point {
    die "Timeline->add_point() expected HASH as parameter" unless scalar(@_) % 2 == 1;

    my ( $self, %data ) = @_;

    %data = $self->_lowercase_keys(%data);
    $self->_required_keys( 'add_point', \%data, (qw/start label/) );
    $self->_valid_keys( 'add_point', \%data, (qw/start label group id/) );

    $data{type} = 'point';
    $data{end}  = $data{start};

    $self->_add_to_pool(%data);
}

sub window {
    my ( $self, %data ) = @_;

    # Default values for our parameters

    $self->{_window_start} = undef;
    $self->{_window_end}   = undef;

    $self->{_window_start_in} = undef;
    $self->{_window_end_in}   = undef;

    $self->{_window_span} = undef;

    $self->{_window_callback} = undef;

    %data = $self->_lowercase_keys(%data);
    $self->_valid_keys( 'window', \%data, (qw/start end start_in end_in span callback/) );

    # Additional validation

    if ( $data{span} ) {
        die "Timeline->window() 'span' can only be defined with a 'start' and 'end'" unless $data{start} and $data{end};
    }

    if ( $data{callback} ) {
        die "Timeline->window() 'callback' can only be a CODE reference" unless ref( $data{callback} ) eq 'CODE';
    }

    foreach my $key ( keys %data ) {
        $self->{"_window_$key"} = $data{$key};
    }
}

sub data {
    my ($self) = @_;

    die "Timeline->data() takes no arguments" if scalar(@_) != 1;

    # Set the start and end, this make things easier

    my $start = ( $self->{_window_start} ? $self->{_window_start} : '0000/00/00T00:00:00' );
    my $end   = ( $self->{_window_end}   ? $self->{_window_end}   : '9999/99/99T23:59:59' );

    my @results;

    if ( $self->{_window_start} ) {
        my $x;
        $x->{start}       = $self->{_window_start};
        $x->{start_start} = $self->{_window_start};
        $x->{start_end}   = $self->{_window_start};

        $x->{end}       = $self->{_window_start};
        $x->{end_start} = $self->{_window_start};
        $x->{end_end}   = $self->{_window_start};

        $x->{type} = 'marker';

        push( @results, $x );
    }

    if ( $self->{_window_end} ) {
        my $x;
        $x->{start}       = $self->{_window_end};
        $x->{start_start} = $self->{_window_end};
        $x->{start_end}   = $self->{_window_end};

        $x->{end}       = $self->{_window_end};
        $x->{end_start} = $self->{_window_end};
        $x->{end_end}   = $self->{_window_end};

        $x->{type} = 'marker';

        push( @results, $x );
    }

    foreach my $record ( @{ $self->{_pool} } ) {
        if ( $record->{start} lt $start ) {
            if ( $record->{end} lt $start ) {
                next;
            }
            elsif ( $record->{end} lt $end ) {
                next unless $self->{_window_end_in};
            }
            else {
                next unless $self->{_window_span};
            }
        }
        elsif ( $record->{start} lt $end ) {
            if ( $record->{end} gt $end ) {
                next unless $self->{_window_start_in};
            }
        }
        else {
            next;
        }

        if ( $self->{_window_callback} ) {
            next unless &{ $self->{_window_callback} }($record);
        }

        if ( $record->{start} lt $start ) {
            $record->{start}       = $start;
            $record->{start_start} = $start;
            $record->{start_end}   = $start;
        }

        if ( $record->{end} gt $end ) {
            $record->{end}       = $end;
            $record->{end_start} = $end;
            $record->{end_end}   = $end;
        }

        push( @results, $record );
    }

    return @results;
}

sub _add_to_pool {
    my ( $self, %data ) = @_;

    my @newpool;
    my $todo = 1;

    %data = $self->_set_range( 'start', %data );
    %data = $self->_set_range( 'end',   %data );

    $data{group} = '--unknown--' unless $data{group};

    foreach my $record ( @{ $self->{_pool} } ) {
        if ( $todo and $record->{start} gt $data{start} ) {
            push @newpool, \%data;
            $todo = undef;
        }
        push @newpool, $record;
    }

    push @newpool, \%data if $todo;

    $self->{_pool} = \@newpool;
}

sub _valid_keys {
    my ( $self, $caller, $data, @keys ) = @_;

    my @testkeys = keys %{$data};
    my %validkeys = map { $_ => $_ } @keys;

    foreach my $key (@testkeys) {
        die "Timeline->$caller() invalid key '$key' passed as data" unless $validkeys{$key};
    }

    foreach my $key ( (qw/start end/) ) {
        if ( $data->{$key} ) {
            $data->{$key} = $self->_today() if $data->{$key} eq 'present';
            die "Timeline->$caller() invalid date for '$key'" unless $self->_date_valid( $data->{$key} );
        }
    }

    if ( $data->{start} and $data->{end} ) {
        die "Timeline->$caller() 'start' and 'end' are in the wrong order" if $data->{start} gt $data->{end};
    }
}

sub _date_valid {
    my ( $self, $date ) = @_;

    my ( $date_part, $time_part ) = split( 'T', $date );
    my ( $year,  $month,   $day )     = split( '\/', $date_part );

    ## Check the date first

    $month = '01' unless $month;
    $day   = '01' unless $day;

    return unless $year  =~ m/^\d+$/;
    return unless $month =~ m/^\d+$/;
    return unless $day   =~ m/^\d+$/;

    my $valid;
    eval { $valid = Date::Calc::check_date( $year, $month, $day ); };

    return unless $valid;

    ## Check the optional time part

    if ($time_part) {
		my ( $hours, $minutes, $seconds ) = split( ':',  $time_part );

        return unless 0 <= $hours   and $hours <= 23;
        return unless 0 <= $minutes and $minutes <= 59;
        return unless 0 <= $seconds and $seconds <= 59;
    }

    return 1;
}

sub _required_keys {
    my ( $self, $caller, $data, @keys ) = @_;

    foreach my $key (@keys) {
        die "Timeline->$caller() missing key '$key'" unless $data->{$key};
    }
}

sub _lowercase_keys {
    my ( $self, %data ) = @_;

    my %newdata = map { lc($_) => $data{$_} } keys %data;

    return %newdata;
}

sub _today {
    my ( $year, $month, $day ) = ( localtime() )[ 5, 4, 3 ];

    $year  += 1900;
    $month += 1;

    return sprintf( "%4d/%02d/%02d", $year, $month, $day );
}

sub _set_range {
    my ( $self, $label, %record ) = @_;

    my ( $date_part, $time_part ) = split( 'T', $record{$label} );
    my ( $year,  $month,   $day )     = split( '\/', $date_part );

    if ($day) {
        $record{"${label}_start"} = $date_part;
        $record{"${label}_end"}   = $date_part;
    }
    elsif ($month) {
        $record{"${label}_start"} = "$year/$month/01";
        $record{"${label}_end"} = "$year/$month/" . Date::Calc::Days_in_Month( $year, $month );
    }
    else {
        $record{"${label}_start"} = "$year/01/01";
        $record{"${label}_end"}   = "$year/12/31";
    }

	if($time_part) {
    $record{"${label}_start"} .= "T" . $time_part;
    $record{"${label}_end"}   .= "T" . $time_part;
	}
	else {
    $record{"${label}_start"} .= "T00:00:00";
    $record{"${label}_end"}   .= "T23:59:59";
}

    return %record;
}

1;

=head1 NAME

Graph::Timeline - Render timeline data

=head1 VERSION

This document refers to verion 1.4 of Graph::Timeline, released June 20, 2007

=head1 SYNOPSIS

This class takes a list of events and processes them so that they can be rendered in 
various graphical formats by subclasses of this class.

=head1 DESCRIPTION

=head2 Overview

The purpose of this class is to organise the data that will be used to render a timeline. Events fall into two types.
Intervals, which has a start and an end. For example Albert Einstein was born on 1879/03/14 and died on 1955/04/18, this would be 
stored as an interval. His works were publicly burned by the Nazi's on 1933/05/10 for being 'of un-German spirit', I guess 
being Jewish didn't help either. So this event would be marked as a point.

You feed events into the class using add_interval( ) and add_point( ), then use window( ) to select which events you want to 
render and then call data( ) to get the relevant events. This last bit will be done in the subclass.

=head2 Constructors and initialisation

=over 4

=item new( )

The constructor takes no arguments and just initialises a few basic variables.

=back

=head2 Public methods

=over 4

=item add_interval( HASH )

Inserts an event that has a start and an end point into the list at the corrct position. The
hash contains the following keys, some of which are required.

=over 4

=item start [ REQUIRED ]

The start date for the interval in the for 'YYYY/MM/DD' or the word 'present' which will be converted into todays date.
Dates in the format YYYY will be taken to span YYYY/01/01 until YYYY/12/31 and dates of the format YYYY/MM will span 
YYYY/MM/01 until YYYY/MM/xx where xx is the last day of MM in YYYY.

=item end [ REQUIRED ]

The start end for the interval in the for 'YYYY/MM/DD' or the word 'present' which will be converted into todays date.
Dates in the format YYYY will be taken to span YYYY/01/01 until YYYY/12/31 and dates of the format YYYY/MM will span 
YYYY/MM/01 until YYYY/MM/xx where xx is the last day of MM in YYYY.

=item label [ REQUIRED ]

The text string that will be displayed when the event is rendered

=item id [ OPTIONAL ]

A unique id for the render, Graph::Timeline does not validate this field for uniqueness

=item group [ OPTIONAL ]

A string is used to group related events together, Graph::Timeline does not validate this field

=back

=item add_point( HASH )

The same as add_interval( ) except that the event occurs on just one day and therefore does not require an end date. Interval and point events are rendered differently.

=item window( HASH )

Set up the data to be selected from the event pool. To reset the defaults just call without any parameters.

=over 4

=item start

Select only record that start on or after this date. Takes a valid date or the word 'present' which is 
translated to the current date.

=item end

Select only record that end on or before this date. Takes a valid date or the word 'present' which is 
translated to the current date.

=item start_in

If end is set then include records that start before the end date but ends after the end date.

=item end_in

If start is set then include records that start before the start date but end after the start date.

=item span

If start and end are both set then additionally report events that start before the start date and end
after the end date.

=item callback

A code reference to provide additionaly custom filtering. The callback will be passed a hash reference with the 
following keys: start, end, label, group, id and type ('interval' and 'point').

=back

=item data( )

This returns a list of the events from the pool that got passed the parameters from the window( ) method.

=back

=head2 Private methods

=over 4

=item _add_to_pool

Used to add the event into the pool which is sorted by start date

=item _valid_keys

Validate that the keys supplied in the hash are valid

=item _date_valid

Validate a date

=item _required_keys

Check that the required keys have been supplied in the hash

=item _lowercase_keys

Lowercase the keys in a hash

=item _today

Return todays date for use with the 'present' word

=item _set_range

Set the xxx_start and xxx_end values from the xxx data

=back

=head1 ENVIRONMENT

None

=head1 DIAGNOSTICS

=over 4

=item Timeline->new() takes no arguments

When the constructor is initialised it requires no arguments. This message is given if 
some arguments were supplied.

=item Timeline->add_interval() expected HASH as parameter

The parameter is a hash describing an event

=item Timeline->add_point() expected HASH as parameter

The parameter is a hash describing an event

=item Timeline->window() 'span' can only be defined with a 'start' and 'end'

To define 'span' then you must also define 'start' and 'end'

=item Timeline->window() 'callback' can only be a CODE reference

You must pass a code reference for the callback 

=item Timeline->data() takes no arguments

When the method is called it requires no arguments. This message is given if 
some arguments were supplied.

=item Timeline->add_interval() invalid key '...' passed as data

The only valid keys are 'start', 'end', 'label', 'group' and 'id'. Something else was supplied.

=item Timeline->add_interval() invalid date for '...'

The date supplied for '...' is invalid

=item Timeline->add_interval() 'start' and 'end' are in the wrong order

The values for 'start' and 'end' are in the wrong order

=item Timeline->add_interval() missing key '...'

A required key was not supplied. Required keys are 'start', 'end' and 'label'

=item Timeline->add_point() invalid key '...' passed as data

The only valid keys are 'start', 'label', 'group' and 'id'. Something else was supplied.

=item Timeline->add_point() invalid date for '...'

The date supplied for '...' is invalid

=item Timeline->add_point() missing key '...'

A required key was not supplied. Required keys are 'start' and 'label'

=item Timeline->window() invalid key '...' passed as data

The only valid keys are 'start', 'end', 'start_in', 'end_in', 'span' and 'callback'. Something else was supplied.

=item Timeline->window() invalid date for '...'

The date supplied for '...' is invalid

=item Timeline->window() 'start' and 'end' are in the wrong order

The values for 'start' and 'end' are in the wrong order

=back

=head1 BUGS

None

=head1 FILES

See the Timeline.t file in the test directory

=head1 SEE ALSO

Graph::Timeline::GD - Use GD to render the timeline

=head1 AUTHORS

Peter Hickman (peterhi@ntlworld.com)

=head1 COPYRIGHT

Copyright (c) 2003, Peter Hickman. All rights reserved.

This module is free software. It may be used, redistributed and/or 
modified under the same terms as Perl itself.
